pkg/daemon/ceph/osd/agent.go

```diff
@@ -426,6 +465,96 @@ func (a *OsdAgent) runOSD(context *clusterd.Context, clusterName string, config
+	initialUsage, err := client.GetOSDUsage(context, a.cluster.Name)
+	if err != nil {
+		logger.Warningf("failed to get baseline OSD usage, but will still continue")
+	}
+
+	// first reweight the OSD to be 0.0, which will begin the data migration
+	o, err := client.CrushReweight(context, a.cluster.Name, config.id, 0.0)
+	if err != nil {
+		return fmt.Errorf("failed to reweight osd.%d to 0.0: %+v. %s", config.id, err, o)
+	}
+
+	// mark the OSD as out
+	if err := markOSDOut(context, a.cluster.Name, config.id); err != nil {
+		return fmt.Errorf("failed to mark osd.%d out: %+v", config.id, err)
+	}
+
+	// wait for the OSDs data to be migrated
+	if err := waitForRebalance(context, a.cluster.Name, config.id, initialUsage); err != nil {
+		return fmt.Errorf("failed to wait for cluster rebalancing after removing osd.%d: %+v", config.id, err)
+	}
+
+	// stop the OSD process and remove it from monitoring
+	if proc, ok := a.osdProc[config.id]; ok {
+		if err := proc.Stop(false); err != nil {
+			return fmt.Errorf("failed to stop proc for osd.%d: %+v", config.id, err)
+		}
+	}
+
+	// purge the OSD from the cluster
+	if err := purgeOSD(context, a.cluster.Name, config.id); err != nil {
+		return fmt.Errorf("failed to purge osd.%d from the cluster: %+v", config.id, err)
+	}
+
+	// delete any backups of the OSD filesystem
+	if err := deleteOSDFileSystem(config); err != nil {
+		logger.Warningf("failed to delete osd.%d filesystem, it may need to be cleaned up manually: %+v", config.id, err)
+	}
+
+	// delete the OSD's local storage
+	osdRootDir := getOSDRootDir(config.configRoot, config.id)
+	if err := os.RemoveAll(osdRootDir); err != nil {
+		logger.Warningf("failed to delete osd.%d root dir from %s, it may need to be cleaned up manually: %+v",
+			config.id, osdRootDir, err)
+	}
+
+	return nil
+}
+
+func waitForRebalance(context *clusterd.Context, clusterName string, osdID int, initialUsage *client.OSDUsage) error {
+	if initialUsage != nil {
+		// start a retry loop to wait for rebalancing to start
+		err := util.Retry(20, 5*time.Second, func() error {
+			currUsage, err := client.GetOSDUsage(context, clusterName)
+			if err != nil {
+				return err
+			}
+
+			init := initialUsage.ByID(osdID)
+			curr := currUsage.ByID(osdID)
+
+			if init == nil || curr == nil {
+				return fmt.Errorf("initial OSD usage or current OSD usage for osd.%d not found. init: %+v, curr: %+v",
+					osdID, initialUsage, currUsage)
+			}
+
+			if curr.UsedKB >= init.UsedKB && curr.Pgs >= init.Pgs {
+				return fmt.Errorf("current used space and pg count for osd.%d has not decreased still", osdID)
+			}
+
+			// either the used space or the number of PGs has decreased for the OSD, data rebalancing has started
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// wait until the cluster gets fully rebalanced again
+	err := util.Retry(3000, 30*time.Second, func() error {
```

Comment:
Yes perhaps this could be made more dynamic (and intelligent) in the future.  If we fail now we will retry again in an idempotent way, so that is probably a good start.